osmomath/binary_search.go

```diff
@@ -26,11 +27,47 @@ type ErrTolerance struct {
 	RoundingDir             RoundingDirection
 }
 
+// Comparable interface is a type constraint that allows any comparable types from osmomath package.
+type Comparable interface {
+	BigDec | Dec | Int
+}
+
 // Compare returns if actual is within errTolerance of expected.
 // returns 0 if it is
 // returns 1 if not, and expected > actual.
 // returns -1 if not, and expected < actual
-func (e ErrTolerance) Compare(expected Int, actual Int) int {
+func Compare[T Comparable](errTolerance ErrTolerance, expected T, actual T) int {
```

Comment:
Do you think that the current approach is a better dev UX relative to non-generic methods?

I was originally hinting at the generics with the goal of reducing code duplication.

If we still have all 3 concrete implementations, the benefits of this approach become questionable

Reply:
yeah, after some digging into go's generics I also think it is just not worth it, I think this PR would only complicate the codebase, without many benefits in return

Reply:
Can we instead implement a generic function that has an interface restriction compatible with the `BigDec`, `Dec`, and `Int` APIs and call this function from every method implementation?

Reply:
https://github.com/osmosis-labs/osmosis/pull/6365#discussion_r1321594155 [note 2]

Reply:
This is not what I'm suggesting.

We should be able to make a generic function without relying on typecasting and calling each concrete method. Instead, the generic function's implementation is shared universally by all math types

Reply:
Do all types already implement that interface? I don't see how that is additional complexity. On the contrary, it would drastically reduce complexity by letting us have one generalized implementation